core/src/main/java/org/mule/runtime/core/TransformationService.java

```diff
@@ -254,9 +254,10 @@ private MuleMessage transformMessage(final MuleMessage message, final MuleEvent
         }
         else
         {
-            return MuleMessage.builder(message)
+            MuleMessage messagePostTransform = (event != null && event.getMessage() != null) ? event.getMessage() : message;
```

Comment:
add a comment explaining this and link a TODO to then corresponding "transformers cleanup" task